scenario/runtime.py

```diff
@@ -1,34 +1,37 @@
-import dataclasses
-import logging
 import marshal
 import os
 import re
-import sys
 import tempfile
 from contextlib import contextmanager
 from pathlib import Path
-from typing import TYPE_CHECKING, Any, Callable, Dict, Optional, Type, TypeVar, Union
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+)
 
 import yaml
 from ops.framework import _event_regex
-from ops.log import JujuLogHandler
 from ops.storage import SQLiteStorage
 
 from scenario.logger import logger as scenario_logger
 from scenario.ops_main_mock import NoObserverError
 
 if TYPE_CHECKING:
-    from ops.charm import CharmBase
-    from ops.framework import EventBase
     from ops.testing import CharmType
 
-    from scenario.state import Event, State, _CharmSpec
+    from scenario.state import DeferredEvent, Event, State, StoredState, _CharmSpec
 
     _CT = TypeVar("_CT", bound=Type[CharmType])
 
 logger = scenario_logger.getChild("runtime")
-# _stored_state_regex = "(.*)\/(\D+)\[(.*)\]"
-_stored_state_regex = "((?P<owner_path>.*)\/)?(?P<data_type_name>\D+)\[(?P<name>.*)\]"
+_stored_state_regex = r"((?P<owner_path>.*)\/)?(?P<data_type_name>\D+)\[(?P<name>.*)\]"
```

Comment:
mmmh I guess so, I did this because that's how ops does it with `_event_regex`, I assumed there was a reason. In general I rarely see module-level regex parsing. Probably a lazy programming style? I think it's good practice to execute as little code as possible when loading a module.

Reply:
I sort of agree that "it's good practice to execute as little code as possible when loading a module", but I think it's better seen as "imports shouldn't have side effects". Compiling a regex is very fast (60us on my machine to compile that one), doesn't have side effects (other than using a tiny bit of CPU and memory), and it's a very common pattern in Python to do it at module load time to avoid the overhead at runtime. I know the time *can* add up if you're doing tons of them, but with just a few I think it's a fine pattern.

We do it that way most of the time in `ops` (`_event_regex` is an odd exception). The Python stdlib source does it commonly too, for example [html.parser](https://github.com/python/cpython/blob/7e5ba351e0702bdff50ff5755082ef2cc5ace381/Lib/html/parser.py#L21), [configparser](https://github.com/python/cpython/blob/7e5ba351e0702bdff50ff5755082ef2cc5ace381/Lib/configparser.py#L1334), and [doctest](https://github.com/python/cpython/blob/7e5ba351e0702bdff50ff5755082ef2cc5ace381/Lib/doctest.py#L589) (and there are many more example).

Reply:
Timing reference:

```
$ python3 -m timeit -s 'import re' 're.compile(r"((?P<owner_path>.*)\/)?(?P<data_type_name>\D+)\[(?P<name>.*)\]"); re.purge()'
5000 loops, best of 5: 64.5 usec per loop
```